### PR TITLE
rdf-ext and @rdfjs/environment and more

### PR DESCRIPTION
**Pull request by tpluscode**
I started this and an update of `rdf-ext` to version 2.0 which has been heavily modified to use ES modules and the new `@rdfjs/environment` package and its factory extensibility model. It also adds some new packages: `@rdfjs/traverser` and `@rdfjs/prefix-map` and brings others up to speed regarding their exported factory types

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an npm package, match the name. If not, do not conflict with the name of an npm package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [x] `tslint.json` [should contain](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#linter-tslintjson) `{ "extends": "@definitelytyped/dtslint/dt.json" }`, and no additional rules.
- [x] `tsconfig.json` [should have](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#tsconfigjson) `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

If changing an existing definition:
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

### CODE REVIEW

**Comment by jeswr** [comment]
@tpluscode - bump me when the checks are passing and I'll try and take a look

@bergos - I note that these updates are largely against repos you're maintaining would you be open to migrating these repositories to Typescript at any point to avoid the need for PRs like this?

**Comment by tpluscode** [comment]
@jeswr would you be fine if I added you as editor to those types? Won't help speed up this but future PRs

**Comment by jeswr** [comment]
Go for it @tpluscode; I'm fairly sure I'm already an editor on a few of them anyway?

**Comment by tpluscode** [comment]
Yes, you were @jeswr. I added you to the other I touched here. 
@cristianvasquez, I also added you to traverser because I know you're a fan :)

**Comment by DangerBotOSS** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## rdf-ext ([<kbd>unpkg</kbd>](https://unpkg.com/browse/rdf-ext@latest/))
was missing the following properties:
1. ClownfaceFactory
1. DataFactory
1. DatasetFactory
1. Environment
1. FetchFactory

<details>
<summary>as well as these 105 other properties...</summary>
<p>FormatsFactory, NamespaceFactory, PrefixMapFactory, TermMapFactory, TraverserFactory, ClownfaceFactory, DataFactory, DatasetFactory, Environment, FetchFactory, FormatsFactory, NamespaceFactory, PrefixMapFactory, TermMapFactory, TraverserFactory, ClownfaceFactory, DataFactory, DatasetFactory, Environment, FetchFactory, FormatsFactory, NamespaceFactory, PrefixMapFactory, TermMapFactory, TraverserFactory, ClownfaceFactory, DataFactory, DatasetFactory, Environment, FetchFactory, FormatsFactory, NamespaceFactory, PrefixMapFactory, TermMapFactory, TraverserFactory, ClownfaceFactory, DataFactory, DatasetFactory, Environment, FetchFactory, FormatsFactory, NamespaceFactory, PrefixMapFactory, TermMapFactory, TraverserFactory, ClownfaceFactory, DataFactory, DatasetFactory, Environment, FetchFactory, FormatsFactory, NamespaceFactory, PrefixMapFactory, TermMapFactory, TraverserFactory, ClownfaceFactory, DataFactory, DatasetFactory, Environment, FetchFactory, FormatsFactory, NamespaceFactory, PrefixMapFactory, TermMapFactory, TraverserFactory, ClownfaceFactory, DataFactory, DatasetFactory, Environment, FetchFactory, FormatsFactory, NamespaceFactory, PrefixMapFactory, TermMapFactory, TraverserFactory, ClownfaceFactory, DataFactory, DatasetFactory, Environment, FetchFactory, FormatsFactory, NamespaceFactory, PrefixMapFactory, TermMapFactory, TraverserFactory, ClownfaceFactory, DataFactory, DatasetFactory, Environment, FetchFactory, FormatsFactory, NamespaceFactory, PrefixMapFactory, TermMapFactory, TraverserFactory, ClownfaceFactory, DataFactory, DatasetFactory, Environment, FetchFactory, FormatsFactory, NamespaceFactory, PrefixMapFactory, TermMapFactory, TraverserFactory</p>
</details>

<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 5f79ccfdb16601b99157fd7cf019e86bc18f52db
</p>

**Comment by tpluscode** [comment]
OMG, ok, I think that does it. I did not like that @DangerBotOSS comment so I also added the previously missing types for `@rdfjs/score`

I mean, just look how big this PR has become. Sorry 🙈

**Comment by tpluscode** [comment]
Ping @jeswr. Buckle up, because it's gonna be a rough ride.

Let me know if maybe you'd like a pair programming..., uh pair review, session

**Comment by jeswr** [comment]
@tpluscode - will do; but probably won't get a chance to look at this until the weekend so if any of the other code owners want to take a look before then go ahead.

**Comment by bergos** [comment]
> @tpluscode - bump me when the checks are passing and I'll try and take a look
> 
> @bergos - I note that these updates are largely against repos you're maintaining. Would you be open to migrating these repositories to Typescript at any point to avoid the need for PRs like this?

I maintain these libraries in my spare time, and my spare time projects are all in plain JS. So I don't benefit from adding TypeScript (definitions).

But I plan to enable sponsoring. If I have some funding, I would be open to adding TypeScript definitions.

**Comment by jeswr** [comment]
@tpluscode since we are already doing bulk changes can we also migrate from using the deprecated `rdf-js` to `@rdfjs/types`.

File: types/rdfjs__term-set/Factory.d.ts [hunk 1]
```diff
@@ -1,7 +1,9 @@
 import TermSet from '.';
-import { Term } from 'rdf-js';
+import { Term } from '@rdfjs/types';
 
-export default class Factory {
-    static readonly exports: ['termSet'];
```
Comment (jeswr) [inline types/rdfjs__term-set/Factory.d.ts:5]: Is there a reason for removing the readonly?

File: types/rdf-validate-shacl/tsconfig.json [hunk 2]
```diff
@@ -14,7 +15,12 @@
         ],
         "types": [],
         "noEmit": true,
-        "forceConsistentCasingInFileNames": true
+        "forceConsistentCasingInFileNames": true,
+        "paths": {
+            "@rdfjs/*": [
+                "rdfjs__*"
+            ]
+        }
```
Comment (jeswr) [inline types/rdf-validate-shacl/tsconfig.json:23]: Why is this here, does it need to be applied to all packages?

File: types/rdf-validate-shacl/tsconfig.json [hunk 3]
```diff
@@ -2,7 +2,8 @@
     "compilerOptions": {
         "module": "commonjs",
         "lib": [
-            "es6"
+            "es6",
+            "dom"
```
Comment (jeswr) [inline types/rdf-validate-shacl/tsconfig.json:6]: Why was dom added?

File: types/rdf-utils-dataset/resourcesToGraph.d.ts [hunk 4]
```diff
@@ -14,6 +13,5 @@ interface Options<F extends Factory> {
 
 type DatasetOf<F> = F extends RDF.DatasetCoreFactory<any, any, infer D> ? D : never;
 
-declare function resourcesToGraph<F extends Factory = Factory<QuadExt, QuadExt, DatasetExt>>(_input: DatasetIndexed<RDF.BaseQuad, RDF.BaseQuad>, options?: Options<F>): DatasetOf<F>;
-
-export = resourcesToGraph;
+export default function resourcesToGraph<F extends Factory = Factory<QuadExt, QuadExt, DatasetExt>>(_input: RDF.DatasetCore<RDF.BaseQuad, RDF.BaseQuad>, options?: Options<F>): DatasetOf<F>;
+export {};
```
Comment (jeswr) [inline types/rdf-utils-dataset/resourcesToGraph.d.ts:17]: Why not have `DatasetExt` as the generic here?

File: types/rdfjs__data-model/lib/fromTerm.d.ts [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+
+export interface FromTerm<T extends RDF.Term, D extends RDF.DataFactory> {
+    (factory: D, original: T):
+        T extends RDF.Literal ? ReturnType<D['literal']>
+            : T extends RDF.BlankNode ? ReturnType<D['blankNode']>
+                : T extends RDF.NamedNode ? ReturnType<D['namedNode']>
+                    : T extends RDF.Variable ? D['variable'] extends (...arg: any[]) => RDF.Variable ? ReturnType<D['variable']> : never
+                        : T extends RDF.Quad ? ReturnType<D['quad']>
+                            : T extends RDF.DefaultGraph ? ReturnType<D['defaultGraph']>
+                                : never;
```
Comment (jeswr) [inline types/rdfjs__data-model/lib/fromTerm.d.ts:11]: It looks like this will infer `never` if I input something of type `BaseQuad` that does not satisfy the `Quad` type. Persumably it should be `T extends RDF.BaseQuad` instead?

File: types/rdfjs__fetch-lite/Factory.d.ts [hunk 6]
```diff
@@ -0,0 +1,20 @@
+import { BaseQuad, DatasetCore, Quad, Stream } from '@rdfjs/types';
+import { FormatsInit } from '.';
+
+interface RdfFetchResponse<D extends DatasetCore<OutQuad, InQuad>, OutQuad extends BaseQuad = Quad, InQuad extends BaseQuad = OutQuad> extends Response {
+    quadStream(): Promise<Stream<OutQuad>>;
+    dataset(): Promise<D>;
+}
+
+export interface FetchFactory {
+    fetch(url: string, options?: FormatsInit): Promise<RdfFetchResponse<DatasetCore>>;
```
Comment (jeswr) [inline types/rdfjs__fetch-lite/Factory.d.ts:10]: The `clone` method is missing.

File: types/rdfjs__fetch-lite/Factory.d.ts [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+}
+
+interface FetchFactoryCtor {
```
Comment (jeswr) [inline types/rdfjs__fetch-lite/Factory.d.ts:13]: What does `Ctor` mean in the name?

File: types/rdfjs__fetch-lite/Factory.d.ts [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+    new(): FetchFactory;
+    exports: ['fetch'];
```
Comment (jeswr) [inline types/rdfjs__fetch-lite/Factory.d.ts:15]: Why is this here?

File: types/rdfjs__fetch-lite/Factory.d.ts [hunk 9]
```diff
@@ -0,0 +1,20 @@
+import { BaseQuad, DatasetCore, Quad, Stream } from '@rdfjs/types';
+import { FormatsInit } from '.';
+
+interface RdfFetchResponse<D extends DatasetCore<OutQuad, InQuad>, OutQuad extends BaseQuad = Quad, InQuad extends BaseQuad = OutQuad> extends Response {
+    quadStream(): Promise<Stream<OutQuad>>;
+    dataset(): Promise<D>;
```
Comment (jeswr) [inline types/rdfjs__fetch-lite/Factory.d.ts:6]: What about the config function (https://github.com/rdfjs-base/fetch-lite/blob/425f5083e7df0a97b7fb5d82c0c616ae5aac5969/Factory.js#L15-L17)?

File: types/rdfjs__namespace/Factory.d.ts [hunk 10]
```diff
@@ -1,6 +1,15 @@
...
 
-export default class Factory {
-    static exports: ['namespace'];
-    namespace(baseIRI: string): NamespaceBuilder;
+export interface NamespaceFactory {
+    // eslint-disable-next-line no-unnecessary-generics
+    namespace<TermNames extends string = any>(baseIRI: string): NamespaceBuilder<TermNames>;
 }
+
+interface NamespaceFactoryCtor {
```
Comment (jeswr) [inline types/rdfjs__namespace/Factory.d.ts:8]: Ditto on `Ctor`

File: types/rdfjs__namespace/Factory.d.ts [hunk 11]
```diff
@@ -1,6 +1,15 @@
 import { NamespaceBuilder } from '.';
 
-export default class Factory {
-    static exports: ['namespace'];
-    namespace(baseIRI: string): NamespaceBuilder;
+export interface NamespaceFactory {
+    // eslint-disable-next-line no-unnecessary-generics
+    namespace<TermNames extends string = any>(baseIRI: string): NamespaceBuilder<TermNames>;
```
Comment (jeswr) [inline types/rdfjs__namespace/Factory.d.ts:5]: Why exactly do you want to make this a generic? Also why is it defaulted to `any` rather than `string`?

File: types/rdfjs__prefix-map/tsconfig.json [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+        "forceConsistentCasingInFileNames": true,
+        "paths": {
+            "@rdfjs/*": [
+                "rdfjs__*"
+            ]
+        }
+    },
+    "files": [
+        "index.d.ts",
+        "rdfjs__prefix-map-tests.ts"
```
Comment (jeswr) [inline types/rdfjs__prefix-map/tsconfig.json:26]: Why is a test included in the files?

File: types/rdfjs__score/Factory.d.ts [hunk 13]
```diff
@@ -0,0 +1,11 @@
+import * as score from '.';
+
+export interface ScoreFactory {
+    score: typeof score;
+}
+
+interface FactoryCtor {
```
Comment (jeswr) [inline types/rdfjs__score/Factory.d.ts:7]: Ditto with `ctor`

**Review by jeswr — COMMENTED** [review]
This is a first pass review of the packages that I'm more familiar with

Questions:
 - Does rdf-ext need changes to the package.json? Same for @rdfjs/data-model.
 
 
What I've not done on each package:
 - I have not checked the tsconfigs/tslints; I'm taking your word for it on these
 - I have not checked test coverage; 
 - I have not checked test files
 - I have not checked version numbers
 - I have not checked for unchanged `rdf-js` usage
 
 I've skipped
  - hydra-box
  - rdf-ext
  - rdfjs__environment
   - rdfjs__score
   - rdfjs__traverser

File: types/rdfjs__data-model/lib/fromTerm.d.ts (hunk 5 above)
Comment (tpluscode) [reply types/rdfjs__data-model/lib/fromTerm.d.ts:11]: Good observation. Unfortunately, now that tried to improve the tests, it appears that it always returns `BaseQuad`... 🙄

File: types/rdf-validate-shacl/tsconfig.json (hunk 2 above)
Comment (tpluscode) [reply types/rdf-validate-shacl/tsconfig.json:23]: Without this DefinitelyTyped cannot resolve scoped packages. Compare with [any random types](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/rebass__forms/tsconfig.json#L17-L20)

File: types/rdf-validate-shacl/tsconfig.json (hunk 3 above)
Comment (tpluscode) [reply types/rdf-validate-shacl/tsconfig.json:6]: Because these tests now indirectly refer to `@rdfjs/fetch-lite` which uses types from global `fetch` interface which is still only defined in the browser env

File: types/rdf-utils-dataset/resourcesToGraph.d.ts (hunk 4 above)
Comment (tpluscode) [reply types/rdf-utils-dataset/resourcesToGraph.d.ts:17]: I cannot answer as I'm not the original author of these types

File: types/rdfjs__data-model/lib/fromTerm.d.ts (hunk 5 above)
Comment (tpluscode) [reply types/rdfjs__data-model/lib/fromTerm.d.ts:11]: Ok, sorry, some GitHub fluke did not allow me to edit my reply.  The factory itself has `BaseQuad` as the generic arg so guess it's correct.

**Comment by tpluscode** [comment]
> Does rdf-ext need changes to the package.json? Same for @rdfjs/data-model.

You refer to the difference in version range?

```diff
      "@rdfjs/types": "^1.0.1"
      "@rdfjs/types": "*"
```

I used star everywhere as `rdf-js` which also do not restrict the version. I kept the existing version in `@rdfjs/data-model` which already used the recommended types package

**Comment by jakebailey** [comment]
So from a packaging point of view, the new packages look fine to me. I can't really review the accuracy of the types; are there any remaining concerns here?

**Comment by tpluscode** [comment]
I don't expect them to be perfect but subsequent fixes would have to come once we try using these types.

Nothing to add from me

**Review by jakebailey — APPROVED** [review]

**Comment by tpluscode** [comment]
Ready to merge
